### ISSUE TITLE
API: inconsistent behavior of `pos_label` across metrics

### ISSUE THREAD

**Issue by cakedev0** [probabl-ai/skore#2592]
### What would you like to say?

While playing around with `pos_label` in various cases for different metrics, I found some behaviors a bit surprising, as mentioned here https://github.com/probabl-ai/skore/pull/2588#issuecomment-4048720725. Here is a table that tries to give a complete view of the various behaviors. Legend:

- **labels column**: labels present in `y`
- **pos_label column**: `pos_label` passed to the `EstimatorReport`
- **other columns**: metrics. Cells in those columns describe what happened when called for the given `labels` and `pos_label`.

| labels    | pos_label   | precision                       | precision(average="binary")          | roc_auc     | brier_score     | roc*                       |
|:----------|:------------|:--------------------------------|:-------------------------------------|:------------|:----------------|:---------------------------|
| [0,1]     | unset       | all classes                     | error: pos_label unset               | scalar      | scalar          | pos_label=1                |
| [0,1]     | fixed (1)   | pos_label=1                     | pos_label=1                          | scalar      | scalar          | pos_label=1                |
| ["A","B"] | unset       | all classes                     | error: pos_label unset               | scalar      | scalar          | error: pos_label required  |
| ["A","B"] | fixed (B)   | pos_label=B                     | pos_label=B                          | scalar      | scalar          | pos_label=B                |
| [0,1,2]   | unset       | all classes                     | error: multiclass + average='binary' | all classes | attribute error | all classes                |
| [0,1,2]   | fixed (2)   | all classes (warning: pos_label ignored) | error: multiclass + average='binary' | all classes | attribute error | all classes                |

\* behavior is the same for the `confusion_matrix` and `precision_recall` displays.

### What surprises me

- Setting `pos_label` seems to have no visible effect for multiclass, I think.
- `roc` / `precision_recall` / `confusion_matrix` plots silently choose a `pos_label` when it is not set and labels are `{0, 1}` or `{-1, 1}`, but crash otherwise.
- `precision(average="binary")` does not seem to bring any feature compared to simply `precision()`.
- Brier score is not supported for multiclass.

### What I propose

I'm not sure yet, this needs discussion. Some ideas:

- **For multiclass**: raise an error if a `pos_label` is set, or use it in some meaningful way?
- **For binary**:
  - enforce picking a `pos_label` (automatically picking `1` if labels are `{0, 1}` or `{-1, 1}`, and requiring it to be set explicitly otherwise)
  - **OR** remove the `average="binary"` option from `precision`/`recall` and avoid silently picking a label / crashing for string labels in the confusion matrix.

**Comment by cakedev0** [comment]
cc @jeromedockes

**Comment by glemaitre** [comment]
I think that having consistency makes sense. I'm just wondering about the multiclass where raising might be too much restrictive in comparison to scikit-learn. I would not try to have a change of behaviour in this case.

**Comment by GaetandeCast** [comment]
I'm ok with forcing the user to choose a value for `pos_label` when labels are not `{0,1}` or `{-1,1}`, it will make things easier to assume in the code that `pos_label` has been set. I don't know if we should be silent when picking `1` when it's relevant though.

**Comment by cakedev0** [comment]
@glemaitre are you ok with silently ignoring `pos_label` for the multi-class case?

**Comment by jeromedockes** [comment]
> assume in the code that pos_label has been set

but not for multiclass right?

**Comment by GaetandeCast** [comment]
>but not for multiclass right?

Yes for binary classif.

Since we never make use of `pos_label` in multiclass (I think), I'm in favor of raising an error when it is set, although I'm not against silently ignoring it. Middle ground may be to warn that it has been set but that it won't have any effect.

**Comment by cakedev0** [comment]
Given the discussions here and offline with @glemaitre I'm moving forward with:
- For multiclass: raise an error if a `pos_label` is set
- For binary: enforce picking a `pos_label` (automatically picking 1 if labels are {0, 1} or {-1, 1}, and requiring it to be set explicitly otherwise)

Plus removing `average="binary"` from `precision` and `recall` as this will be redundant.

**Comment by cakedev0** [comment]
[edit: **this plan is outdated**, but the final goal remains the same]

Big change of plan, based on discussions offline, here is the new plan:

## 1: remove `pos_label` from the init and clean-up the API

- remove `pos_label` from the report's init (and remove it from attributes)
- re-introduce `pos_label` optional kwarg in: `precision`, `recall`, `roc` curve, `precision_recall` curve, `confusion_matrix`. This kwarg will only be accepted for binary classification tasks.
- **never automatically pick a `pos_label`: this makes the API inconsistent.**
- remove `average="binary"` option for precision & recall: this is redundant with passing a `pos_label`

How it plays out with the hub:
- we'll always send data for both classes 
- the hub will implement a toggle to select the `pos_label`

Detailed target behavior of metrics/plots:

| Metric             | Multiclass              | Binary (no `pos_label`) | Binary (with `pos_label`)       |
|--------------------|--------------------------|--------------------------|----------------------------------|
| accuracy             | scalar | scalar                   | n/a                           |
| ROC AUC            | dict (OVR)               | scalar                   | n/a                           |
| Precision / Recall | dict (all classes)       | dict (all classes)       | scalar                           |
| ROC curve          | $n_c$ curves             | 2 curves                 | one curve                        |
| PR curve           | $n_c$ curves             | all classes              | one curve                        |
| Confusion matrix   | $n_c$ x $n_c$              | 2x2                      | 2x2 (top-left = TP?)              |

Notes:
- confusion matrix in the binary classification case:
  - Except in scikit-learn, top-left always seems to be the true positive. Should be align on that?
  - We might want to differentiate more the plots with vs without `pos_label` so that we don't mislead people thinking that something is the TP when it's not really (because no `pos_label` was given). This could be a follow-up PR/issue.
- `brier_score`: I think adding support for the multiclass case is easy and not really related to this issue.

## 2: add back `pos_label` in the init in the form of an optional `default_pos_label`

In metrics/plots, change `pos_label` default from `None` (all classes) to `<DEFAULT>` which will use `self.default_pos_label` (that will be `None` by default)

What's nice with this change: it's backward compatible in terms of API. So we don't have to implement it right now. It's just a convenience that mimics the toggle button on the hub's side.

**Comment by cakedev0** [comment]
Current state of things now that #2663 is merged:
- `pos_label` behavior is consistent
- `pos_label` isn't required to build a binary classification report, nor to compute `pos_label`-dependent metrics/displays (like the precision or the PR curve): we now display for both.

What remains to be done would be to allow passing `pos_label` to `pos_label`-dependent metrics/displays, it would overwrite the report's pos_label. This can be done in several steps/PRs and will effectively make the report's pos_label a `default_pos_label` (we'll see at the end if we want to rename it or not).

We don't need to remove `pos_label` from the init to do that, so let's keep it so we don't make any breaking changes: we're just adding kwargs to metrics/displays 👌

**Comment by cakedev0** [comment]
> What remains to be done would be to allow passing pos_label to pos_label-dependent metrics/displays, it would overwrite the report's pos_label.

Since `report.metrics.add` was merged, we're good on the metrics' side, you can now do things like this:

```Python
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from skore import evaluate
from sklearn.metrics import make_scorer, precision_score

X, y = make_classification(random_state=0)
report = evaluate(LogisticRegression(), X, y, splitter=0.2, pos_label=1)
report.metrics.add(make_scorer(precision_score, average=None, pos_label=None), name="Precision (both)")
report.metrics.add(make_scorer(precision_score, average="binary", pos_label=0), name="Precision (0)")

report.metrics.summarize().frame()
```

The only thing that remains is the confusion matrix.
